README.md
#ptomulik-macro

[![Build Status](https://travis-ci.org/ptomulik/puppet-macro.png?branch=master)](https://travis-ci.org/ptomulik/puppet-macro)
[![Coverage Status](https://coveralls.io/repos/ptomulik/puppet-macro/badge.png)](https://coveralls.io/r/ptomulik/puppet-macro)
[![Code Climate](https://codeclimate.com/github/ptomulik/puppet-macro.png)](https://codeclimate.com/github/ptomulik/puppet-macro)

####<a id="table-of-contents"></a>Table of Contents

1. [Overview](#overview)
2. [Module Description](#module-description)
3. [Usage](#usage)
   * [Example 1: Defining macro in ruby code](#example-1-defining-macro-in-ruby-code)
   * [Example 2: Invoking macro in puppet manifest](#example-2-invoking-macro-in-puppet-manifest)
   * [Example 3: Macro with parameters](#example-3-macro-with-parameters)
   * [Example 4: Variable number of parameters](#example-4-variable-number-of-parameters)
   * [Example 5: Default parameters](#example-5-default-parameters)
   * [Example 6: Invoking macro from macro](#example-6-invoking-macro-from-macro)
   * [Example 7: Using variables](#example-7-using-variables)
   * [Example 8: Building dependencies between parameters](#example-8-building-dependencies-between-parameters)
4. [Reference](#reference)
   * [Function Reference](#function-reference)
   * [API Reference](#api-reference)
5. [Limitations](#limitations)

##<a id="overview"></a>Overview

Puppet parser macros.

[[Table of Contents](#table-of-contents)]

##<a id="module-description"></a>Module Description

With this functionality module developers may define named macros and evaluate
them in manifests. This works similarly to parser
[functions](http://docs.puppetlabs.com/guides/custom_functions.html) but
implementing a macro is a little bit easier. Also, macros may use
"hierarchical" names in `foo::bar::geez` form, so a module developer may easily
establish one-to-one correspondence between macro names and class/define
parameters.

The main reason for this module being developed is exemplified in
[Example 8](#example-8-building-dependencies-between-parameters).

[[Table of Contents](#table-of-contents)]

##<a id="usage"></a>Usage

###<a id="example-1-defining-macro-in-ruby-code"></a>Example 1: Defining macro in ruby code

To define a macro named **foo::bar** in a module write a file
named *lib/puppet/macros/foo/bar.rb*:

```ruby
# lib/puppet/macros/foo/bar.rb
Puppet::Macros.newmacro 'foo::bar' do ||
  'macro foo::bar'
end
```

The above macro simply returns the `'macro foo::bar'` string. Note the empty
argument list `||`. This enforces strict arity checking (zero arguments) on
ruby **1.8**. Without `||` the block is assumed to accept arbitrary number of
arguments on **1.8** and no arguments on **1.9**. 

[[Table of Contents](#table-of-contents)]

###<a id="example-2-invoking-macro-in-puppet-manifest"></a>Example 2: Invoking macro in puppet manifest

Nothing simpler than:

```puppet
$foo_bar = determine('foo::bar')
notify { foo_bar: message => "determine('foo::bar') -> ${foo_bar}" }
```

If you don't need the value returned by macro, then you may invoke macro as
a statement:

```puppet
invoke('foo::bar')
```

[[Table of Contents](#table-of-contents)]

###<a id="example-3-macro-with-parameters"></a>Example 3: Macro with parameters

Let's define macro `sum2` which adds two integers:

```ruby
# lib/puppet/macros/sum2.rb
Puppet::Macros.newmacro 'sum2' do |x,y|
  Integer(x) + Integer(y)
end
```

Now `sum2` may be used as follows:

```puppet
$sum = determine('sum2', 1, 2)
notify { sum: message => "determine('sum2',1,2) -> ${sum}" }
```

[[Table of Contents](#table-of-contents)]

###<a id="example-4-variable-number-of-parameters"></a>Example 4: Variable number of parameters

Let's redefine macro from [Example 3](#example-3-macro-with-parameters) to
accept arbitrary number of parameters:

```ruby
# lib/puppet/macros/sum.rb
Puppet::Macros.newmacro 'sum' do |*args|
  args.map{|x| Integer(x)}.reduce(0,:+)
end
```

Now, few experiments:

```puppet
$zero = determine('sum')
$one = determine('sum',1)
$three = determine('sum',1,2)
notify { zero: message => "determine('sum') -> ${zero}" }
notify { one: message => "determine('sum',1) -> ${one}" }
notify { three: message => "determine('sum',1,2) -> ${three}" }
```

[[Table of Contents](#table-of-contents)]

###<a id="example-5-default-parameters"></a>Example 5: Default parameters

Default parameters work only with ruby **1.9+**. If you don't care about
compatibility with ruby **1.8**, you may define a macro with default parameters
in the usual way:

```ruby
# lib/puppet/macros/puppet/config/content.rb
Puppet::Macros.newmacro 'puppet::config::content' do |file='/etc/puppet/puppet.conf'|
  File.read(file)
end
```

Now you may use it with:

```puppet
$content = determine('puppet::config::content')
notify { content: message => $content }
```

or

```puppet
$content = determine('puppet::config::content','/usr/local/etc/puppet/puppet.conf')
notify { content: message => $content }
```

If you need the same for ruby *1.8*, here is a workaround (note that the
caller i.e the [determine](#determine) function, will check the minimum arity,
so we only check the maximum):

```ruby
# lib/puppet/macros/puppet/config/content.rb
Puppet::Macros.newmacro 'puppet::config::content' do |*args|
  if args.size > 1
    raise Puppet::ParseError, "Wrong number of arguments (#{args.size} for maximum 1)"
  end
  args << '/etc/puppet/puppet.conf' if args.size < 1
  File.read(args[0])
end
```

[[Table of Contents](#table-of-contents)]

###<a id="example-6-invoking-macro-from-macro"></a>Example 6: Invoking macro from macro

You may invoke macro using `call_macro` method:

```ruby
# lib/puppet/macros/bar.rb
Puppet::Macros.newmacro 'bar' do
  call_macro('foo::bar')
end
```

The first argument to `call_macro` is the name of the macro to be invoked, the
second (if present) is an array of arguments to be passed to macro.

You may alternatively use function interface, but this isn't the recommended
way (you may receive misleading exception messages in case you mess up with
arguments to macro).

```ruby
# lib/puppet/macros/bar.rb
Puppet::Macros.newmacro 'bar' do
  function_determine(['foo::bar'])
end
```

If you test any of the above with the following puppet code:

```puppet
$bar = determine('bar')
notify { bar: message => "determine('bar') -> ${bar}" }
```

then the following notice would appear on output:

```console
Notice: determine('bar') -> macro foo::bar
```

Obviously the above text is the result of `foo::bar` macro defined in
[Example 1](#example-1-defining-macro-in-ruby-code).

[[Table of Contents](#table-of-contents)]

###<a id="example-7-using-variables"></a>Example 7: Using variables

You may access puppet variables, for example `$::osfamily` (fact).  The
following example determines default location of apache configs for operating
system running on slave:

```ruby
# lib/puppet/macros/apache/conf_dir.rb
Puppet::Macros.newmacro 'apache::conf_dir' do
  case os = lookupvar("::osfamily")
  when /FreeBSD/; '/usr/local/etc/apache22'
  when /Debian/; '/usr/etc/apache2'
  else
    raise Puppet::Error, "unsupported osfamily #{os.inspect}"
  end
end
```

```puppet
$apache_conf_dir = determine('apache::conf_dir')
notify { apache_conf_dir: message => "determine('apache::conf_dir') -> ${apache_conf_dir}" }
```

[[Table of Contents](#table-of-contents)]

###<a id="example-8-building-dependencies-between-parameters"></a>Example 8: Building dependencies between parameters

Macros may be used to inter-depend parameters of defined types or classes. In
other words, if one parameter is altered by user, others should be adjusted
automatically, unless user specify them explicitly. For example, we may have a
defined type `testmodule::foo` with two parameters `$a` and `$b` and we want
`$b` to depend on `$a`.  So, we may define macros `testmodule::foo::a` and
`testmodule::foo::b`, and `testmodule::foo::b` may accept `$a` as an argument:

```ruby
# lib/puppet/macros/testmodule/foo/a.rb
Puppet::Macros.newmacro 'testmodule::foo::a' do |a|
    pp2r(a) ? a : 'default a'
end
```

```ruby
# lib/puppet/macros/testmodule/foo/b.rb
Puppet::Macros.newmacro 'testmodule::foo::b' do |b, a|
    pp2r(b) ? b : "default b for a=#{a.inspect}"
end
```

Then, if we split `testmodule::foo` into actual implementation (let say
`testmodule::impl::foo`) and a wrapper (let's call it simply `testmodule::foo`),
the job may be finished as follows:

```puppet
# manifests/impl/foo.pp
define testmodule::impl::foo($a, $b) {
  notify{$title: message => "${title}: a=\'${a}\', b=\'${b}\'"}
}
```

```puppet
# manifests/foo.pp
define testmodule::foo($a = undef, $b = undef)
{
  $_a = determine('testmodule::foo::a', $a)
  $_b = determine('testmodule::foo::b', $b, $_a)
  testmodule::impl::foo{"$title":
    a => $_a,
    b => $_b
  }
}
```

Now, the following manifest

```console
puppet apply --modulepath $(pwd) <<!
testmodule::foo {defaults: }
testmodule::foo {custom_a: a => 'custom a' }
testmodule::foo {custom_b: b => 'custom b' }
testmodule::foo {custom_a_and_b: a => 'custom a', b => 'custom b' }
testmodule::foo {other: }
Testmodule::Foo[other] { a => 'other default a' }
!
```

would output these lines:

```console
Notice: defaults: a='default a', b='default b for a="default a"'
Notice: custom_a: a='custom a', b='default b for a="custom a"'
Notice: custom_b: a='default a', b='custom b'
Notice: custom_a_and_b: a='custom a', b='custom b'
Notice: other: a='other default a', b='default b for a="other default a"'
```

[[Table of Contents](#table-of-contents)]

##<a id="reference"></a>Reference

###<a id="function-reference"></a>Function reference

####<a id="index-of-functions"></a>Index of functions:

* [determine](#determine)
* [invoke](#invoke)

####<a id="determine"></a>determine
Determine value of a macro.

This function ivokes a macro defined with `Puppet::Macros.newmacro`
method and returns its value. The function takes macro name as first
argument and macro parameters as the rest of arguments. The number of
arguments provided by user is validated against macro's arity.

*Example*:

Let say, you have defined the following macro in
*puppet/macros/sum.rb*:

    # puppet/macros/sum.rb
    Puppet::Macros.newmacro 'sum' do |x,y|
      Integer(x) + Integer(y)
    end

You may then invoke the macro from puppet as follows:

    $three = determine('sum',1,2) # -> 3

- *Type*: rvalue

[[Index of functions](#index-of-functions)|[Table of Contents](#table-of-contents)]

####<a id="invoke"></a>invoke
Invoke macro as a statement.

This function ivokes a macro defined with `Puppet::Macros.newmacro`
method. The function takes macro name as first argument and macro parameters
as the rest of arguments. The number of arguments provided by user is
validated against macro's arity.

*Example*:

Let say, you have defined the following macro in
*puppet/macros/print.rb*:

    # puppet/macros/pring.rb
    Puppet::Macros.newmacro 'print' do |msg|
      print msg
    end

You may then invoke the macro from puppet as follows:

    invoke('print',"hello world!\\n")

- *Type*: statement

[[Index of functions](#index-of-functions)|[Table of Contents](#table-of-contents)]

###<a id="api-reference"></a>API Reference

API reference may be generated with

```console
bundle exec rake yard
```

The generated documentation goes to `doc/` directory. Note that this works only
under ruby >= 1.9.

The API documentation is also available
[online](http://rdoc.info/github/ptomulik/puppet-macro/).


[[Table of Contents](#table-of-contents)]

##Limitations

* Currently there is no possibility to define macro in puppet manifests, that is
  we only can define macro using ruby and use it in ruby or puppet. I believe
  this functionality may implemented as an additional parser function (call it
  `macro`) and it should work with the help of puppet lambdas, which are
  available in future parser.
* Currently there is no way to store and auto-generate documentation for macros.
  It should work similarly as for functions but its not implemented at the
  moment. This may be added in future versions.

[[Table of Contents](#table-of-contents)]


CHANGELOG
2014-03-12 Pawel Tomulik <ptomulik@meil.pw.edu.pl>
  * release 0.1.1
  * fix path expansion in spec_helper_integration.rb
  * corrected example 8 in README.md
2014-01-02 Pawel Tomulik <ptomulik@meil.pw.edu.pl>
  * fixed monkey patching of Puppet::Util::Autoload.expand
2014-01-30 Pawel Tomulik <ptomulik@meil.pw.edu.pl>
  * small corrections to API docs
  * configured yardoc to use github flawored markup
  * release 0.1.0
  * moved macros from puppet/parser/macros to puppet/macros
2014-01-28 Pawel Tomulik <ptomulik@meil.pw.edu.pl>
  * use pp2r() in Example 8
  * release 0.0.18
  * added fr2r method to Scope
  * release 0.0.17
  * fixed bug with missing Autoload.expand() method
  * release 0.0.16
  * added pp2r method to Scope
2014-01-27 Pawel Tomulik <ptomulik@meil.pw.edu.pl>
  * release 0.0.15
  * added call_macro to scope and fixed lambda binding issues
  * release 0.0.14
  * revised example 6
  * release 0.0.13
  * fixed specs for ruby 2.0
  * prevent adding duplicates in $LOAD_PATH
  * added specs for validators
  * release 0.0.12
  * revised Example 8
  * release 0.0.11
  * fixed bug with arity validation faililng on 1.9+
  * cleanup yard documentation
  * release 0.0.10
  * switched back from lambdas to blocks in documentation and specs
  * fixed typo in README.md
  * release 0.0.9
  * revised README.md
  * release 0.0.8
  * revised README.md
  * revised documentation for functions
  * release 0.0.7
  * removed workaround for travis bug
  * refactoring
2014-01-26 Pawel Tomulik <ptomulik@meil.pw.edu.pl>
  * release 0.0.6
  * added few "integration" tests
  * included puppet 3.3 and 3.4 on travis
  * included puppet 2.7 on travis
  * optimized 'invoke' function for better code coverage
  * release 0.0.5
  * documented functions determine and invoke
  * revised README.md
  * revised Example 8
  * release 0.0.4
  * reworked Example 8
2014-01-25 Pawel Tomulik <ptomulik@meil.pw.edu.pl>
  * release 0.0.3
  * revised README.md
  * release 0.0.2
  * added specs to better cover newmacro, determine and invoke functions
  * release 0.0.1
  * corrected Example 6 in README.md
  * initial commit


spec/spec_helper.rb
if RUBY_VERSION >= "1.9"
  require 'coveralls'
  Coveralls.wear! do
    add_filter 'spec/'
  end
end

require 'puppetlabs_spec_helper/module_spec_helper'


spec/spec_helper_integration.rb
require 'spec_helper.rb'

RSpec.configuration.module_path.split(File::PATH_SEPARATOR).each do |mp|
  testmod = File.expand_path(File.join(mp,'testmodule','lib'))
  if testmod =~ /#{Regexp.escape(File.join('spec','fixtures','modules'))}/
    $LOAD_PATH.unshift(testmod) unless $LOAD_PATH.include?(testmod)
  end
end


lib/puppet/parser/functions/invoke.rb
dir = File.expand_path(File.join(File.dirname(__FILE__),'..','..','..'))
$LOAD_PATH.unshift(dir) unless $LOAD_PATH.include?(dir)
require 'puppet/macros'
module Puppet::Parser::Functions
  newfunction(:invoke, :type => :statement, :doc => <<-EOT
  Invoke macro as a statement.

  This function ivokes a macro defined with `Puppet::Macros.newmacro`
  method. The function takes macro name as first argument and macro parameters
  as the rest of arguments. The number of arguments provided by user is
  validated against macro's arity.

  *Example*:

  Let say, you have defined the following macro in
  *puppet/macros/print.rb*:

      # puppet/macros/pring.rb
      Puppet::Macros.newmacro 'print' do |msg|
        print msg
      end

  You may then invoke the macro from puppet as follows:

      invoke('print',"hello world!\\n")
  EOT
  ) do |args|
    Puppet::Macros.call_macro_from_func(self,:invoke,args)
  end
end


lib/puppet/parser/functions/determine.rb
dir = File.expand_path(File.join(File.dirname(__FILE__),'..','..','..'))
$LOAD_PATH.unshift(dir) unless $LOAD_PATH.include?(dir)
require 'puppet/macros'
module Puppet::Parser::Functions
  newfunction(:determine, :type => :rvalue, :doc => <<-EOT
  Determine value of a macro.

  This function ivokes a macro defined with `Puppet::Macros.newmacro`
  method and returns its value. The function takes macro name as first
  argument and macro parameters as the rest of arguments. The number of
  arguments provided by user is validated against macro's arity.

  *Example*:

  Let say, you have defined the following macro in
  *puppet/macros/sum.rb*:

      # puppet/macros/sum.rb
      Puppet::Macros.newmacro 'sum' do |x,y|
        Integer(x) + Integer(y)
      end

  You may then invoke the macro from puppet as follows:

      $three = determine('sum',1,2) # -> 3
  EOT
  ) do |args|
    Puppet::Macros.call_macro_from_func(self,:determine,args)
  end
end


lib/puppet/macros.rb
require 'puppet'
require 'puppet/util/autoload'

# Monkey patches to Scope, so we have convenient access to some methods from
# macros.
class Puppet::Parser::Scope
  # Call macro registered with {Puppet::Macros.newmacro}
  #
  # @param name [String] macro name, e.g. `'foo::bar'`,
  # @param args [Array] arguments to be provided to the macro,
  # @param options [Hash] additional options, see {Puppet::Macros.call_macro}
  # @param env [Puppet::Node::Environment] puppet environment,
  # @return the result of macro evaluation
  def call_macro(name, args = [], options = {}, env = Puppet::Macros.default_environment)
    # FIXME: I believe I should use environment that is bound to scope, not the
    # Macros.default_environment, but at the moment I have no idea where to
    # find it
    Puppet::Macros.call_macro(self,name,args,options,env)
  end

  # Convert puppet value to ruby.
  #
  # This converts empty strings and :undefs to nil.
  #
  # @param v [String|Symbol] value to be converted
  # @return converted value
  def pp2r(v)
    (v.equal?(:undef) or v == '') ? nil : v
  end

  # Convert value from facter to ruby.
  #
  # This converts empty strings and :undefined symbol to nil.
  #
  # @param v [String|Symbol] value to be converted
  # @return converted value
  def fr2r(v)
    (v.equal?(:undefined) or v == '') ? nil : v
  end
end

module Puppet::Macros; end

# Utility module for {Puppet::Macros}
module Puppet::Macros::DefaultEnvironment
  # This tries to ensure compatibility with different versions of Puppet
  # @api private
  if Puppet.respond_to?(:lookup)
    def default_environment
      Puppet.lookup(:current_environment)
    end
  else
    begin
      require 'puppet/context'
      def default_environment
        Puppet::Context.lookup(:current_environment)
      end
    rescue LoadError
      begin
        require 'puppet/node/environment'
        def default_environment
          Puppet::Node::Environment.current
        end
      rescue LoadError
        def default_environment
          nil
        end
      end
    end
  end
end

# Utility module for {Puppet::Macros}
module Puppet::Macros::Validation
  # Validate name
  #
  # @param name [Object] the name to be validated
  # @raise [ArgumentError]
  # @api private
  def validate_name(name, errclass = ArgumentError)
    unless valid_name?(name)
      raise errclass, "Invalid macro name #{name.inspect}"
    end
  end

  # @api private
  def macro_arities_by_parameters(macro)
    arg_kinds = macro.parameters.map{|kind,name| kind}
    [arg_kinds.count(:req), arg_kinds.include?(:rest) ? :inf : arg_kinds.size]
  end

  # @api private
  def macro_arities_by_arity(macro)
    arity = macro.arity
    (arity>=0) ? [arity, arity] : [arity.abs-1, :inf]
  end

  # @api private
  def macro_arities(macro)
    # Using macro.parameters is far more reliable than macro.arity, but
    # parameters are missing in ruby<=1.8.
    if macro.respond_to?(:parameters)
      macro_arities_by_parameters(macro)
    else
      macro_arities_by_arity(macro)
    end
  end

  # @api private
  def check_macro_arity(macro, macro_args, errclass = ArgumentError)
    min_arity, max_arity = macro_arities(macro)
    argn = macro_args.size
    if min_arity == max_arity
      if argn != min_arity
        raise errclass, "Wrong number of arguments (#{argn} for #{min_arity})"
      end
    elsif argn < min_arity
      raise errclass, "Wrong number of arguments (#{argn} for minimum #{min_arity})"
    elsif (not max_arity.equal?(:inf)) and (argn > max_arity)
      raise errclass, "Wrong number of arguments (#{argn} for maximum #{max_arity})"
    end
  end
end

module Puppet::Macros::ToLambda
  def to_lambda(block)
    if Puppet::Util::Package.versioncmp(RUBY_VERSION,"1.9") >=0 
      unless block.lambda?
        # This code is based on: https://github.com/schneems/proc_to_lambda
        # See also https://github.com/schneems/proc_to_lambda/issues/1
        if RUBY_ENGINE && RUBY_ENGINE == "jruby"
          lambda(&block)
        else
          Object.new.define_singleton_method(:_, &block).to_proc
        end
      end
    else
      block
    end
  end
end

# This object keeps track of macros defined within a single puppet
# environment. Existing hashes (macros for existing environments) may be
# retrieved with {macros} method.
module Puppet::Macros

  class << self
    include DefaultEnvironment
    include Validation
    include ToLambda

    # Regular expession used to validate macro names.
    MACRO_NAME_RE =  /^[a-z_][a-z0-9_]*(?:::[a-z_][a-z0-9_]*)*$/

    # Check whether **name** is a valid macro name.
    #
    # @param name a name to be validated
    # @return [Boolean] `true` if the **name** is valid, or `false` otherwise
    # @api private
    def valid_name?(name)
      name.is_a?(String) and MACRO_NAME_RE.match(name)
    end

    # Define new preprocessor macro.
    #
    # A preprocessor macro is a callable object, which can be executed from
    # within a puppet manifest.
    #
    # **Example:**
    #
    #  Definition:
    #
    #  ```ruby
    #  # puppet/macros/apache/package.rb
    #  Puppet::Parser.Macro.newmacro 'apache::package' do
    #    setcode do
    #      case os = fact(:osfamily)
    #      when 'FreeBSD'
    #        'www/apache22'
    #      when 'Debian'
    #        'apache2'
    #      else
    #        raise Puppet::Error, "#{os} is not supported"
    #      end
    #    end
    #  end
    #  ```
    #
    # Usage:
    #
    #  ```puppet
    #  # manifest.pp
    #  $package = determine('apache::package')
    #  ````
    #
    # @param name [String] macro name
    # @param options [Hash] additional options
    # @param block [Proc] macro definition
    #
    # @option options environment [Puppet::Node::Environment] an environment
    #   this macro belongs to, defautls to `default_environment`
    def newmacro(name,options={},&block)
      env = options[:environment] || default_environment
      Puppet.debug "overwritting macro #{name}" if macro(name,env,false)
      validate_name(name)
      macros(env)[name] = to_lambda(block)
    end

    # Get a hash of registered macros.
    #
    # @param env [Puppet::Node::Environment] puppet environment
    # @return [Hash] a hash with registered parser macros
    # @api private
    def macros(env = default_environment)
      @macros ||= {}
      @macros[env] ||= {}
    end

    # Retrieve single macro from {macros}
    #
    # @param name [String] macro name,
    # @param env [Puppet::Node::Environment] puppet environment,
    def macro(name, env = default_environment, auto = true)
      root = Puppet::Node::Environment.root
      macros(env)[name] || macros(root)[name] || (auto ? load(name,env) : nil)
    end

    # Accessor for singleton autoloader
    # @api private
    def autoloader
      unless @autoloader
        # Patched autoloader whith 'loadall' searching recursivelly
        @autoloader = Puppet::Util::Autoload.new(
          self, "puppet/macros", :wrap => false
        )
        class << @autoloader
          def loadall
            self.class.loadall(File.join(@path,"**"))
          end
          def files_to_load
            self.class.files_to_load(File.join(@path,"**"))
          end
          unless instance_methods.include?(:expand) or instance_methods.include?('expand')
            # Fix for puppet 2.7, which does not have the Autoload.expand
            # method
            def expand(name)
              ::File.join(@path, name.to_s)
            end
          end
        end
      end
      @autoloader
    end

    # Load single macro from file
    #
    # @param name [String] macro name, e.g. `'foo::bar'`,
    # @param env [Puppet::Node::Environment] puppet environment,
    # @return [Macro|nil]
    # @api private
    def load(name, env = default_environment)
      # This block autoloads appropriate file each time a missing macro is
      # requested from hash.
      path = name.split('::').join('/')
      load_from_file(name, path, env)
    end

    # Load single file possibly containing macro definition.
    #
    # @param name [String] name of the macro to be loaded
    # @param path [String] path to macro file, relative to
    #   **puppet/macros** and without `.rb` suffix,
    # @param env [Puppet::Node::Environment] puppet environment,
    # @return [Macro|nil]
    # @api private
    def load_from_file(name, path, env = default_environment)
      if autoloader.load(path, env)
        # the autoloaded code should add its macro to macros
        unless m = self.macro(name,env,false)
          Puppet.debug("#{autoloader.expand(path).inspect} loaded but it " +
            "didn't define macro #{name.inspect}")
        end
        m
      else
        Puppet.debug("could not autoload #{autoloader.expand(path).inspect}")
        nil
      end
    end

    # Autoload all existing macro definitions in current environment
    #
    # @param env [Puppet::Node::Environment] puppet environment,
    # @return [Array] an array of loaded files
    def loadall
      autoloader.loadall
    end

    # @api private
    def get_macro(name, env = default_environment, errclass = Puppet::Error)
      unless macro = self.macro(name,env)
        raise errclass, "Undefined macro #{name}"
      end
      macro
    end

    # Fix error messages to indicate number of arguments to parser function
    # instead of the number of arguments to macro and prepend the function
    # name.
    # @param func [Symbol|String] function name,
    # @param msg [String] original message from callee,
    # @param n [Integer] number of arguments shifted from function's arglist,
    # @return [String] fixed message
    # @api private
    def fix_error_msg(func, msg, n = 0)
      re = /^Wrong number of arguments \(([0-9]+) for (minimum |maximum )?([0-9]+)\)$/
      if m = re.match(msg)
        msg = "Wrong number of arguments (#{n+Integer(m.captures[0])} " +
          "for #{m.captures[1]}#{n+Integer(m.captures[2])})"
      end
      "#{func}(): #{msg}"
    end


    # Call a macro.
    #
    # @param scope [Puppet::Parser::Scope] scope of the calling function
    # @param name [String] name of the macro to be invoked
    # @param args [Array] arguments to be provided to teh macro
    # @param options [Hash] additional options
    # @param env [Puppet::Node::Environment] environment
    # @option options :a_err [Class] an exception to be raised when argument
    #   validation fails, defaults to **ArgumentError**
    # @option options :l_err [Class] an exception to be raised when macro
    #   lookup fails, defaults to **Puppet::Error**
    # @return the value of macro (result of evaluation)
    def call_macro(scope, name, args, options = {}, env = default_environment)
      validate_name(name, options[:a_err] || ArgumentError)
      macro = get_macro(name, env, options[:l_err] || Puppet::Error)
      check_macro_arity(macro, args, options[:a_err] || ArgumentError)
      scope.instance_exec(*args,&macro)
    end

    # Call a macro from parser function.
    #
    # This method is dedicated to be called from a parser function. It's used
    # by `determine` and `invoke`, but may be used by other custom functions as
    # well. This method checks the arguments and in case of validation error
    # raises Puppet::ParseError with appropriate message. If there is error
    # in number of arguments to macro, the exception message will reflect the
    # number of arguments to function, not the macro.
    #
    #
    # @param scope [Puppet::Parser::Scope] scope of the calling function
    # @param func_name [Symbol|String] name of the calling function
    # @param func_args [Array] arguments, as provided to calling function
    # @param n [Integer] number of extra arguments shifted from function's
    #   argument list
    # @param env [Puppet::Node::Environment] environment
    # @return the value of macro (result of evaluation)
    def call_macro_from_func(scope, func_name, func_args, n = 0, env = default_environment)
      begin
        if(func_args.size == 0)
          raise Puppet::ParseError, "Wrong number of arguments (0) - missing macro name"
        end
        options = { :a_err => Puppet::ParseError, :l_err => Puppet::ParseError }
        call_macro(scope, func_args[0], func_args[1..-1], options, env)
      rescue Puppet::ParseError => err
        msg = fix_error_msg(func_name, err.message, 1+n)
        raise Puppet::ParseError, msg, err.backtrace
      end
    end
  end
end
